api/server/server.go

```diff
@@ -177,26 +180,35 @@ func (s *Server) Start(ctx context.Context) {
 }
 
 func (s *Server) startGears(ctx context.Context) {
+	// By default it serves on :8080 unless a
+	// PORT environment variable was defined.
+	listen := fmt.Sprintf(":%d", viper.GetInt(EnvPort))
+	listener, err := net.Listen("tcp", listen)
+	if err != nil {
+		logrus.WithError(err).Fatalln("Failed to serve functions API.")
+	}
+	logrus.Infof("Serving Functions API on address `%s`", listen)
+
 	svr := &supervisor.Supervisor{
 		MaxRestarts: supervisor.AlwaysRestart,
 		Log: func(msg interface{}) {
 			logrus.Debug("supervisor: ", msg)
 		},
 	}
 
-	// By default it serves on :8080 unless a
-	// PORT environment variable was defined.
 	svr.AddFunc(func(ctx context.Context) {
-		go s.Router.Run()
+		go http.Serve(listener, s.Router)
 		<-ctx.Done()
 	})
 
 	svr.AddFunc(func(ctx context.Context) {
-		runner.StartWorkers(ctx, s.Runner, s.tasks)
+		go runner.RunAsyncRunner(ctx, s.apiURL, s.tasks, s.Runner)
```

Comment:
if I understand correctly, `runner.RunAsyncRunner` is blocking already. No need for `go` plus reading the `ctx.Done()` channel.